glean-core/src/common_metric_data.rs

```diff
@@ -72,7 +75,52 @@ pub struct CommonMetricData {
     /// metric to be recorded to, dynamic labels are stored in the specific
     /// label so that we can validate them when the Glean singleton is
     /// available.
-    pub dynamic_label: Option<String>,
+    pub dynamic_label: Option<DynamicLabelType>,
+}
+
+// TODO FIX ERRORS. Similar errors see for all variants, not sure how to specify this in UniFFI idl:
+/*
+error[E0769]: tuple variant `Self::Label` written as struct variant
+   --> <redacted>/glean/target/debug/build/glean-core-7b31c5607feb2bbe/out/glean.uniffi.rs:125:1
+    |
+125 | #[::uniffi::udl_derive(Enum)]
+    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+    |
+    = note: this error originates in the attribute macro `::uniffi::udl_derive` (in Nightly builds, run with -Z macro-backtrace for more info)
+
+*/
+
+#[derive(Debug, Clone, Deserialize, Serialize, MallocSizeOf)]
+pub enum DynamicLabelType {
+    Label(String),
+    KeyOnly(String),
+    CategoryOnly(String),
+    KeyAndCategory(String),
+}
+
+impl Default for DynamicLabelType {
+    fn default() -> Self {
+        Self::Label(String::new())
+    }
+}
+
+impl Deref for DynamicLabelType {
+    type Target = str;
+
+    fn deref(&self) -> &Self::Target {
+        match self {
+            DynamicLabelType::Label(label) => label,
+            DynamicLabelType::KeyOnly(key) => key,
+            DynamicLabelType::CategoryOnly(category) => category,
+            DynamicLabelType::KeyAndCategory(key_and_category) => key_and_category,
+        }
+    }
+}
+
+impl Display for DynamicLabelType {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}", self.to_string())
+    }
 }
```

Comment:
I don't think we need display (plus this potentially double allocates due to the `to_string()`). Leave it out.

```suggestion
```